### PR TITLE
Add logging on startup/shutdown

### PR DESCRIPTION
**Pull request by erikjohnston**
This is so we can tell what is going on when things are taking a while to start up.

The main change here is to ensure that transactions that are created during startup get correctly logged like normal transactions.

### CODE REVIEW

**Comment by clokep** [comment]
@erikjohnston What's with the removals of the `convert_param_style`? Was that made obsolete at some point?

**Comment by erikjohnston** [comment]
> @erikjohnston What's with the removals of the `convert_param_style`? Was that made obsolete at some point?

The `LoggingTransaction` does it for you, which is why they're only there when using the native cursors.

File: synapse/storage/database.py [hunk 1]
```diff
@@ -247,6 +301,12 @@ def _do_execute(self, func, sql: str, *args: Any) -> None:
     def close(self) -> None:
         self.txn.close()
 
+    def __enter__(self) -> "LoggingTransaction":
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
```
Comment (clokep) [inline synapse/storage/database.py:308]: Did we never previously used `LoggingTransaction` as a context manager even though `Transaction` can be used that way?

File: synapse/storage/database.py [hunk 2]
```diff
@@ -584,7 +642,10 @@ def inner_func(conn, *args, **kwargs):
                     logger.debug("Reconnecting closed database connection")
                     conn.reconnect()
 
-                return func(conn, *args, **kwargs)
+                db_conn = LoggingDatabaseConnection(
+                    conn, self.engine, "runWithConnection"
+                )
+                return func(db_conn, *args, **kwargs)
```
Comment (clokep) [inline synapse/storage/database.py:648]: Is this ensuring another location uses `LoggingDatabaseConnection` or is it necessary because of the refactoring? I'm having a bit of trouble following why this change was done.

File: synapse/storage/database.py [hunk 3]
```diff
@@ -1642,8 +1703,6 @@ def get_cache_dict(
             "limit": limit,
         }
 
-        sql = self.engine.convert_param_style(sql)
-
         txn = db_conn.cursor()
```
Comment (clokep) [inline synapse/storage/database.py:1706]: The type of `db_conn` should now be `LoggingDatabaseConnection`, I think?

File: synapse/storage/databases/main/schema/delta/56/unique_user_filter_index.py [hunk 4]
```diff
@@ -46,7 +48,4 @@ def run_create(cur, database_engine, *args, **kwargs):
         select_clause,
     )
 
-    if isinstance(database_engine, PostgresEngine):
-        cur.execute(sql)
-    else:
-        cur.executescript(sql)
+    execute_statements_from_stream(cur, StringIO(sql))
```
Comment (clokep) [inline synapse/storage/databases/main/schema/delta/56/unique_user_filter_index.py:51]: Why this change? It seems potentially inefficient?

**Review by clokep — COMMENTED** [review]
Seems pretty reasonable, my tl;dr of this change is: everything now uses `LoggingDatabaseConnection`, thus everything now uses `LoggingTransaction`, thus we can remove a bunch of manual code.

Does that seem accurate?

File: synapse/storage/database.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/storage/database.py:308]: Exactly.

File: synapse/storage/database.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/storage/database.py:648]: This just ensures that whenever we have use a connection after a call to `runWithConnection` it is a logging connection, which a) helps ensure that more SQL statements get logged and b) necessary to make the tests work  as they call `runWithConnection` and then call things expecting a logging connection a lot.

File: synapse/storage/databases/main/schema/delta/56/unique_user_filter_index.py (hunk 4 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/schema/delta/56/unique_user_filter_index.py:51]: This is a delta so it doesn't really matter if its inefficient. We do this because we don't expose an `executescript` on `LoggingTranasction`, and implementing it would be a bit of a faff.

**Review by clokep — APPROVED** [review]

**Comment by erikjohnston** [comment]
> Seems pretty reasonable, my tl;dr of this change is: everything now uses `LoggingDatabaseConnection`, thus everything now uses `LoggingTransaction`, thus we can remove a bunch of manual code.
> 
> Does that seem accurate?

Yup :slightly_smiling_face:
